src/libraries/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfo.cs

```diff
@@ -250,6 +250,34 @@ private static double GetNumericValueNoBoundsCheck(uint codePoint)
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToUpper(uint codePoint)
```

Comment:
Both this method and `ToLower` will work on _any_ Unicode code point, not just BMP code points. This means that you can rewrite `SurrogateCasing` in terms of this function, removing the special-casing for Osage and other languages. So this method should return `uint`, not `char`.

You'll also want to update `Rune.ToUpperInvariant` and `Rune.ToLowerInvariant` to call into this routine if invariant mode is active. Otherwise those results will begin to diverge from `TextInfo.ToUpper`, etc.

Finally, the call to `GetCategoryCasingTableOffsetNoBoundsChecks` can buffer overrun if the input is not within the valid Unicode code point range (`U+0000 .. U+10FFFF`). Take a look through the rest of this file, and you'll see that the methods which take `(uint codePoint)` as an argument are all private, and the only entry points to them are internal methods which validate their arguments before forwarding it to these methods. (If the internal method accepts a `char`, it's a 16-bit value so is already limited to `U+0000 .. U+FFFF`, which is a valid code point; and if the internal method accepts a `Rune`, the type safety system already guarantees it's a legal code point.) So be careful when you expose these methods to arbitrary internal callers.

Here's an example of internal methods which validate arguments before passing them to private (non-validating) helpers:

https://github.com/dotnet/runtime/blob/4c92aef2b08f9c4374c520e7e664a44f1ad8ce56/src/libraries/System.Private.CoreLib/src/System/Globalization/CharUnicodeInfo.cs#L261-L274